## Import Libraries

In [1]:
import pandas as pd, numpy as np
from plotly.subplots import make_subplots
import plotly.graph_objects as go

#Set float number just have two decimals
pd.set_option('display.float_format','{:.2f}' .format)
#Set option display max columns
pd.options.display.max_columns=999

## Load and Concat Dataset

In [2]:
df1 = pd.read_csv('Premier_league_18_19.csv')
df2 = pd.read_csv('Premier_league_19_20.csv')
df3 = pd.read_csv('Premier_league_20_21.csv')

In [3]:
df = pd.concat([df1,df2,df3])
df.head()

date          home_team        away_team  home_score  away_score  \
0  08/12/2018            Arsenal  Manchester City           0           2   
1  08/11/2018    AFC Bournemouth     Cardiff City           2           0   
2  08/11/2018             Fulham   Crystal Palace           0           2   
3  08/11/2018  Huddersfield Town          Chelsea           0           3   
4  08/12/2018          Liverpool  West Ham United           4           0   

   home_possession_%  away_possession_%  home_shots_on_target  \
0              42.00              58.00                  3.00   
1              62.90              37.10                  4.00   
2              66.30              33.70                  6.00   
3              37.20              62.80                  1.00   
4              64.80              35.20                  8.00   

   away_shots_on_target  home_shots  away_shots  home_touches  away_touches  \
0                  8.00        9.00       17.00        593.00        734.00   
1                  1.00       12.00       10.00        711.00        483.00   
2                 10.00       15.00       12.00        893.00        564.00   
3                  4.00        6.00       13.00        550.00        856.00   
4                  2.00       18.00        5.00        843.00        551.00   

   home_passes  away_passes  home_tackles  away_tackles  home_clearances  \
0       401.00       555.00         22.00          6.00            23.00   
1       502.00       287.00         11.00         22.00            37.00   
2       672.00       347.00         25.00         21.00            22.00   
3       372.00       658.00         25.00         12.00            13.00   
4       673.00       360.00         19.00         23.00             8.00   

   away_clearances  home_corners  away_corners  home_offsides  away_offsides  \
0             9.00          2.00          9.00           7.00           2.00   
1            30.00          7.00          4.00           0.00           2.00   
2            30.00          5.00          5.00           2.00           3.00   
3            21.00          2.00          5.00           2.00           1.00   
4            20.00          5.00          4.00           5.00           3.00   

   home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards  \
0               2.00               2.00            0.00            0.00   
1               1.00               1.00            0.00            0.00   
2               1.00               2.00            0.00            0.00   
3               2.00               1.00            0.00            0.00   
4               1.00               2.00            0.00            0.00   

   home_fouls_conceded  away_fouls_conceded  
0                11.00                14.00  
1                11.00                 9.00  
2                 9.00                11.00  
3                 9.00                 8.00  
4                14.00                 9.00

 ## Data Manipulation

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 379
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   date                  1140 non-null   object 
 1   home_team             1140 non-null   object 
 2   away_team             1140 non-null   object 
 3   home_score            1140 non-null   int64  
 4   away_score            1140 non-null   int64  
 5   home_possession_%     1140 non-null   float64
 6   away_possession_%     1140 non-null   float64
 7   home_shots_on_target  1140 non-null   float64
 8   away_shots_on_target  1140 non-null   float64
 9   home_shots            1140 non-null   float64
 10  away_shots            1140 non-null   float64
 11  home_touches          1140 non-null   float64
 12  away_touches          1140 non-null   float64
 13  home_passes           1140 non-null   float64
 14  away_passes           1140 non-null   float64
 15  home_tackles          

### Handling Datetime Column

In [5]:
#convert 'date' column datatype to datetime
df['date'] = pd.to_datetime(df['date'], format='%m/%d/%Y')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 379
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   date                  1140 non-null   datetime64[ns]
 1   home_team             1140 non-null   object        
 2   away_team             1140 non-null   object        
 3   home_score            1140 non-null   int64         
 4   away_score            1140 non-null   int64         
 5   home_possession_%     1140 non-null   float64       
 6   away_possession_%     1140 non-null   float64       
 7   home_shots_on_target  1140 non-null   float64       
 8   away_shots_on_target  1140 non-null   float64       
 9   home_shots            1140 non-null   float64       
 10  away_shots            1140 non-null   float64       
 11  home_touches          1140 non-null   float64       
 12  away_touches          1140 non-null   float64       
 13  home_passes        

### Sorting Dataframe 

In [6]:
#sorting dataframe by the date
df = df.sort_values(by='date', ascending=True, ignore_index=True)
df

date                home_team                 away_team  \
0    2018-08-11          AFC Bournemouth              Cardiff City   
1    2018-08-11                   Fulham            Crystal Palace   
2    2018-08-11        Huddersfield Town                   Chelsea   
3    2018-08-11        Manchester United            Leicester City   
4    2018-08-11         Newcastle United         Tottenham Hotspur   
...         ...                      ...                       ...   
1135 2021-05-23                   Fulham          Newcastle United   
1136 2021-05-23                  Arsenal  Brighton and Hove Albion   
1137 2021-05-23          West Ham United               Southampton   
1138 2021-05-23             Leeds United      West Bromwich Albion   
1139 2021-05-23  Wolverhampton Wanderers         Manchester United   

      home_score  away_score  home_possession_%  away_possession_%  \
0              2           0              62.90              37.10   
1              0           2              66.30              33.70   
2              0           3              37.20              62.80   
3              2           1              46.30              53.70   
4              1           2              40.40              59.60   
...          ...         ...                ...                ...   
1135           0           2              64.30              35.70   
1136           2           0              53.90              46.10   
1137           3           0              36.90              63.10   
1138           3           1              64.40              35.60   
1139           1           2              57.20              42.80   

      home_shots_on_target  away_shots_on_target  home_shots  away_shots  \
0                     4.00                  1.00       12.00       10.00   
1                     6.00                 10.00       15.00       12.00   
2                     1.00                  4.00        6.00       13.00   
3                     6.00                  4.00        8.00       13.00   
4                     2.00                  5.00       15.00       15.00   
...                    ...                   ...         ...         ...   
1135                  0.00                  4.00       14.00       10.00   
1136                  5.00                  1.00       16.00        5.00   
1137                  7.00                  5.00       14.00       17.00   
1138                  9.00                  5.00       17.00       14.00   
1139                  4.00                  4.00       14.00        9.00   

      home_touches  away_touches  home_passes  away_passes  home_tackles  \
0           711.00        483.00       502.00       287.00         11.00   
1           893.00        564.00       672.00       347.00         25.00   
2           550.00        856.00       372.00       658.00         25.00   
3           697.00        744.00       485.00       543.00         26.00   
4           622.00        809.00       387.00       573.00         15.00   
...            ...           ...          ...          ...           ...   
1135        778.00        517.00       596.00       333.00         18.00   
1136        727.00        645.00       565.00       489.00         14.00   
1137        483.00        715.00       300.00       534.00         12.00   
1138        685.00        465.00       499.00       268.00         21.00   
1139        744.00        634.00       571.00       440.00         19.00   

      away_tackles  home_clearances  away_clearances  home_corners  \
0            22.00            37.00            30.00          7.00   
1            21.00            22.00            30.00          5.00   
2            12.00            13.00            21.00          2.00   
3            14.00            37.00            14.00          2.00   
4            27.00            17.00            43.00          3.00   
...            ...              ...              ...           ...   
1135         

### Dropping, Adding, and Renaming Columns

In [7]:
#Dropping columns that will not be used for analysis
drop_cols = ['home_touches','away_touches', 
             'home_passes', 'away_passes', 
             'home_tackles', 'away_tackles', 
             'home_clearances', 'away_clearances', 
             'home_corners', 'away_corners',
             'home_offsides', 'away_offsides', 
             'home_yellow_cards', 'away_yellow_cards', 
             'home_red_cards', 'away_red_cards', 
             'home_fouls_conceded', 'away_fouls_conceded']
df = df.drop(drop_cols, axis=1)

In [8]:
#Renaming home_score and away_score to home_goals and away_goals
df=df.rename({'home_score':'home_goals', 'away_score':'away_goals'}, axis=1)

#Renaming home_shots_on_target and away_shots_on_target to home_SoT and away_SoT
df=df.rename({'home_shots_on_target':'home_SoT', 'away_shots_on_target':'away_SoT'}, axis=1)

#Renaming home_possession_% and away_possession_% to home_possession and away_possession
df=df.rename({'home_possession_%':'home_possession', 'away_possession_%':'away_possession'}, axis=1)


Adding some variables of team performance metrics to be analyzed:
- Shot on Target % (SoT_rate) = Shots on Target / Shots (Home/Away)
- Quantity Conversion Rate (QuanCR) = Goals / Shots (Home/Away)
- Quality Conversion Rate (QualCR) = Goals / Shots on Target (Home/Away)
- Points per game (ppg) = 3 if a team win, 1 if a match draw, 0 if a team lose
- Win rate (win_rate) = 100 if a team win in a match, 0 if they lose or draw (Home/Away)

In [9]:
df['home_SoT_rate']= df['home_SoT']/df['home_shots']*100
df['away_SoT_rate']= df['away_SoT']/df['away_shots']*100
df['home_QuanCR'] = df['home_goals']/df['home_shots']*100
df['away_QuanCR'] = df['away_goals']/df['away_shots']*100
df['home_QualCR'] = df['home_goals']/df['home_SoT']*100
df['away_QualCR'] = df['away_goals']/df['away_SoT']*100

for i in range(len(df)):
    if df.loc[i,'home_goals']>df.loc[i,'away_goals']:
        df.loc[i,'home_ppg'] = 3
        df.loc[i,'away_ppg'] = 0
    elif df.loc[i,'home_goals']==df.loc[i,'away_goals']:
        df.loc[i,'home_ppg'] = 1
        df.loc[i,'away_ppg'] = 1
    else:
        df.loc[i,'home_ppg'] = 0
        df.loc[i,'away_ppg'] = 3
        
df.loc[df['home_goals']>df['away_goals'], 'home_win_rate'] = 100
df.loc[df['home_goals']<df['away_goals'], 'home_win_rate'] = 0
df.loc[df['home_goals']==df['away_goals'], 'home_win_rate'] = 0

df.loc[df['away_goals']>df['home_goals'], 'away_win_rate'] = 100
df.loc[df['away_goals']<df['home_goals'], 'away_win_rate'] = 0
df.loc[df['away_goals']==df['home_goals'], 'away_win_rate'] = 0
#There will be nan value, a product of 0/0. Replace or fill the na value with 0.
df = df.fillna(0, axis=1)

In [10]:
df.head()

date          home_team          away_team  home_goals  away_goals  \
0 2018-08-11    AFC Bournemouth       Cardiff City           2           0   
1 2018-08-11             Fulham     Crystal Palace           0           2   
2 2018-08-11  Huddersfield Town            Chelsea           0           3   
3 2018-08-11  Manchester United     Leicester City           2           1   
4 2018-08-11   Newcastle United  Tottenham Hotspur           1           2   

   home_possession  away_possession  home_SoT  away_SoT  home_shots  \
0            62.90            37.10      4.00      1.00       12.00   
1            66.30            33.70      6.00     10.00       15.00   
2            37.20            62.80      1.00      4.00        6.00   
3            46.30            53.70      6.00      4.00        8.00   
4            40.40            59.60      2.00      5.00       15.00   

   away_shots  home_SoT_rate  away_SoT_rate  home_QuanCR  away_QuanCR  \
0       10.00          33.33          10.00        16.67         0.00   
1       12.00          40.00          83.33         0.00        16.67   
2       13.00          16.67          30.77         0.00        23.08   
3       13.00          75.00          30.77        25.00         7.69   
4       15.00          13.33          33.33         6.67        13.33   

   home_QualCR  away_QualCR  home_ppg  away_ppg  home_win_rate  away_win_rate  
0        50.00         0.00      3.00      0.00         100.00           0.00  
1         0.00        20.00      0.00      3.00           0.00         100.00  
2         0.00        75.00      0.00      3.00           0.00         100.00  
3        33.33        25.00      3.00      0.00         100.00           0.00  
4        50.00        40.00      0.00      3.00           0.00         100.00

### Pre Attendance Prohibition and Post Attendance Prohibition Data Manipulation

Last matchweek before covid-19 attendance/crowd prohibition: 
   Matchweek 29: 10 March 2020 Leicester vs Aston Villa

In [11]:
#last match before covid-19 suspension
last_idx = df.loc[df['date']=='2020-03-10'].index[0]

In [12]:
#partitioning data before and after covid-19 suspension
pre_covid_data = df[:last_idx+1]
post_covid_data = df[last_idx+1:]

To make analysis from 3 EPL seasons, we only take teams that played in those all 3 seasons and eliminate data of teams that didn't.

2018/2019 relegated teams:
- Fulham
- Cardiff City
- Huddersfield Town

2019/2020 promoted teams:
- Norwich City
- Sheffield Utd
- Aston Villa

2019/2020 relegated teams:
- Bournemouth
- Watford
- Norwich City

2020/2021 promoted teams:
- Leeds Utd
- West Brom
- Fulham

In [13]:
all_teams = list(df['home_team'].unique())
relegated_promoted_teams = ['Fulham', 'Huddersfield Town', 'Cardiff City', 
                  'Norwich City', 'Sheffield United', 'Aston Villa', 
                  'AFC Bournemouth', 'Watford', 'Leeds United', 'West Bromwich Albion']

#stayed_teams = list(set(all_teams)-set(relegated_promoted_teams))
stayed_teams = [team for team in all_teams if team not in relegated_promoted_teams]

In [14]:
'''home_pre_covid_pts = pre_covid_data.groupby(['home_team']).agg(avg_pts_before = ('home_ppg','mean')).loc[stayed_teams].reset_index()
home_post_covid_pts = post_covid_data.groupby(['home_team']).agg(avg_pts_after = ('home_ppg','mean')).loc[stayed_teams].reset_index()
home_df = home_pre_covid_pts.merge(home_post_covid_pts,on='home_team')
home_df'''

"home_pre_covid_pts = pre_covid_data.groupby(['home_team']).agg(avg_pts_before = ('home_ppg','mean')).loc[stayed_teams].reset_index()\nhome_post_covid_pts = post_covid_data.groupby(['home_team']).agg(avg_pts_after = ('home_ppg','mean')).loc[stayed_teams].reset_index()\nhome_df = home_pre_covid_pts.merge(home_post_covid_pts,on='home_team')\nhome_df"

In [15]:
'''away_pre_covid_pts = pre_covid_data.groupby(['away_team']).agg(avg_pts_before = ('away_ppg','mean')).loc[stayed_teams].reset_index()
away_post_covid_pts = post_covid_data.groupby(['away_team']).agg(avg_pts_after = ('away_ppg','mean')).loc[stayed_teams].reset_index()
away_df = away_pre_covid_pts.merge(away_post_covid_pts,on='away_team')
away_df'''

"away_pre_covid_pts = pre_covid_data.groupby(['away_team']).agg(avg_pts_before = ('away_ppg','mean')).loc[stayed_teams].reset_index()\naway_post_covid_pts = post_covid_data.groupby(['away_team']).agg(avg_pts_after = ('away_ppg','mean')).loc[stayed_teams].reset_index()\naway_df = away_pre_covid_pts.merge(away_post_covid_pts,on='away_team')\naway_df"

#### Create home and away statistics function

In [16]:
def home_stats(stat, team, summary=True):
    #aggregating team playing at home before and after attendance prohibition
    home_pre_covid_stat = pre_covid_data.groupby(['home_team']).agg({stat:'mean'}).loc[team].reset_index()
    home_pre_covid_stat = home_pre_covid_stat.rename(columns={stat:'{}_before'.format(stat)})
    home_post_covid_stat = post_covid_data.groupby(['home_team']).agg({stat:'mean'}).loc[team].reset_index()
    home_post_covid_stat = home_post_covid_stat.rename(columns={stat:'{}_after'.format(stat)})
    
    #merge df of before and after attendance prohibition
    home_df = home_pre_covid_stat.merge(home_post_covid_stat,on='home_team')
    
    #extract the 'changes' of team performance before and after attendance prohibition when playing at home
    home_df['changes'] = home_df.iloc[:,2] - home_df.iloc[:,1]
    
    #rate_of_change_% = changes/stats before attendance prohibition*100
    home_df['rate_of_change_%'] = home_df.iloc[:,3]/home_df.iloc[:,1]*100
    
    #sort dataframe by 'changes' ascending
    home_df = home_df.sort_values(by='changes', ascending=True).reset_index(drop=True)
    
    #Summary of average stats changes of all teams
    if summary==True:
        
        effect_categories= []
        #Summary of attendance prohibition effect to team stats
        for i in home_df['rate_of_change_%']:
            if i>5:
                effect_categories.append('positively affected')
            elif i<-5:
                effect_categories.append('negatively affected')
            else:
                effect_categories.append('relatively unaffected')
        effect_categories=pd.Series(effect_categories).value_counts()
        
        #Summary of average stats before attendance prohibition
        avg_stat_before = round(sum(home_df[stat+'_before'])/len(home_df[stat+'_before']),2)
        #Summary of average stats after attendance prohibition
        avg_stat_after = round(sum(home_df[stat+'_after'])/len(home_df[stat+'_after']),2)
        #Summary of average stat changes
        avg_stat_changes = round(sum(home_df['changes'])/len(home_df['changes']),2)
        #Summary of average stats changes of all teams
        avg_rate_of_change = round(avg_stat_changes/avg_stat_before*100,2)
        
        #Print the summary text
        print('Before attendance prohibition, the average {} is {}.'.format(stat, avg_stat_before, end=''))
        print('After attendance prohibition, the average {} is {}.'.format(stat, avg_stat_after, end=''))
        print('The average {} changes is {}.'.format(stat, avg_stat_changes, end=''))
        print('The average {} rate_of_change is {}%.'.format(stat, avg_rate_of_change, end=''))
        print('There are {} clubs are {}, {} {}, and {} {}.'.format(effect_categories.values[0], effect_categories.index[0],
                                                                   effect_categories.values[1], effect_categories.index[1],
                                                                   effect_categories.values[2], effect_categories.index[2]))
        
    return home_df

In [17]:
def away_stats(stat,team, summary=True):
    #aggregating team playing at away match before and after attendance prohibition
    away_pre_covid_stat = pre_covid_data.groupby(['away_team']).agg({stat:'mean'}).loc[team].reset_index()
    away_pre_covid_stat = away_pre_covid_stat.rename(columns={stat:'{}_before'.format(stat)})
    away_post_covid_stat = post_covid_data.groupby(['away_team']).agg({stat:'mean'}).loc[team].reset_index()
    away_post_covid_stat = away_post_covid_stat.rename(columns={stat:'{}_after'.format(stat)})
    
    #merge df of before and after lockdown
    away_df = away_pre_covid_stat.merge(away_post_covid_stat,on='away_team')
    
    #extract the 'changes' of team performance before and after attendance prohibition when playing at away stadium
    away_df['changes'] = away_df.iloc[:,2] - away_df.iloc[:,1]
    
    #rate_of_change_% = changes/stats before crowd attendance prohibition*100
    away_df['rate_of_change_%'] = away_df.iloc[:,3]/away_df.iloc[:,1]*100
    
    #sort dataframe by 'changes' ascending
    away_df = away_df.sort_values(by='changes', ascending=True).reset_index(drop=True)
    
    #Summary of average stats changes of all teams
    if summary==True:
        
        effect_categories= []
        #Summary of attendance prohibition effect to team stats
        for i in away_df['rate_of_change_%']:
            if i>5:
                effect_categories.append('positively affected')
            elif i<-5:
                effect_categories.append('negatively affected')
            else:
                effect_categories.append('relatively unaffected')
        effect_categories=pd.Series(effect_categories).value_counts()
        
        #Summary of average stats before attendance prohibition
        avg_stat_before = round(sum(away_df[stat+'_before'])/len(away_df[stat+'_before']),2)
        #Summary of average stats after attendance prohibition
        avg_stat_after = round(sum(away_df[stat+'_after'])/len(away_df[stat+'_after']),2)
        #Summary of average stat changes
        avg_stat_changes = round(sum(away_df['changes'])/len(away_df['changes']),2)
        #Summary of average stats changes of all teams
        avg_rate_of_change = round(avg_stat_changes/avg_stat_before*100,2)
        
        #Print the summary text
        print('Before attendance prohibition, the average {} is {}.'.format(stat, avg_stat_before, end=''))
        print('After attendance prohibition, the average {} is {}.'.format(stat, avg_stat_after, end=''))
        print('The average {} changes is {}.'.format(stat, avg_stat_changes, end=''))
        print('The average {} rate_of_change is {}%.'.format(stat, avg_rate_of_change, end=''))
        print('There are {} clubs are {}, {} {}, and {} {}.'.format(effect_categories.values[0], effect_categories.index[0],
                                                                   effect_categories.values[1], effect_categories.index[1],
                                                                   effect_categories.values[2], effect_categories.index[2]))
    
    return away_df

We must categorized the changes of stats after attendance prohibition to conclude if a team stats are affected by it or not. The categories will be positively affected, negatively affected, and relatively unaffected. However we don't want to categorize a -0.01 change of stats into negatively affected because the change is not meaningful. Hence, we will use a threshold value to define an effect meaningful or not.  

With the threshold of 5%, we categorized rate_of_change_% of each team into 3 categories to conclude if a team stats affected by attendance prohibition or not because we:
1. **Positively affected** if the rate_of_change_% >5%
2. **Negatively affected** if the rate_of_change_% < -5%
3. **Relatively unaffected** if the rate_of_change_% between <= -5% and >= 5%

The stats are categorized into 2 categories:
1. Result stats: win rate, points per game (ppg)
2. Performance stats: goals per game (gpg), possession, shots, shots on target (SoT), shots on target percentage(SoT_%), Quality Goal Conversion Rate (QualCR), Quantity Goal Conversion Rate (QuanCR)


#### RESULT STATS

##### 1. Win Rate

In [18]:
home_win_rate = home_stats('home_win_rate', stayed_teams)
home_win_rate

Before attendance prohibition, the average home_win_rate is 50.96.
After attendance prohibition, the average home_win_rate is 43.65.
The average home_win_rate changes is -7.31.
The average home_win_rate rate_of_change is -14.34%.
There are 9 clubs are negatively affected, 4 relatively unaffected, and 2 positively affected.


home_team  home_win_rate_before  home_win_rate_after  \
0                  Liverpool                 94.12                56.52   
1                    Everton                 51.52                29.17   
2                    Burnley                 41.18                21.74   
3                    Arsenal                 61.76                47.83   
4           Newcastle United                 39.39                29.17   
5   Brighton and Hove Albion                 30.30                20.83   
6            Manchester City                 84.38                76.00   
7             Crystal Palace                 32.35                26.09   
8          Manchester United                 52.94                47.83   
9          Tottenham Hotspur                 60.61                58.33   
10            Leicester City                 50.00                47.83   
11   Wolverhampton Wanderers                 44.12                43.48   
12                   Chelsea                 55.88                56.52   
13           West Ham United                 39.39                50.00   
14               Southampton                 26.47                43.48   

    changes  rate_of_change_%  
0    -37.60            -39.95  
1    -22.35            -43.38  
2    -19.44            -47.20  
3    -13.94            -22.57  
4    -10.23            -25.96  
5     -9.47            -31.25  
6     -8.38             -9.93  
7     -6.27            -19.37  
8     -5.12             -9.66  
9     -2.27             -3.75  
10    -2.17             -4.35  
11    -0.64             -1.45  
12     0.64              1.14  
13    10.61             26.92  
14    17.01             64.25

Information obtained from home_win_rate dataframe:
- 12 out of 15  clubs had dropped win rate (9 clubs dropped significantly) with average of -14.34% rate_of_change after crowd suspension when playing at home ground
- Liverpool suffers the most from crowd suspension, with  -37.60 win rate dropped when playing at home.

In [19]:
away_stats('away_win_rate', stayed_teams)

Before attendance prohibition, the average away_win_rate is 37.23.
After attendance prohibition, the average away_win_rate is 43.67.
The average away_win_rate changes is 6.44.
The average away_win_rate rate_of_change is 17.3%.
There are 8 clubs are positively affected, 4 relatively unaffected, and 3 negatively affected.


away_team  away_win_rate_before  away_win_rate_after  \
0                  Liverpool                 75.76                50.00   
1             Crystal Palace                 39.39                29.17   
2    Wolverhampton Wanderers                 33.33                29.17   
3          Tottenham Hotspur                 41.18                39.13   
4                Southampton                 30.30                29.17   
5                    Chelsea                 48.48                50.00   
6            Manchester City                 67.65                69.57   
7             Leicester City                 42.42                45.83   
8           Newcastle United                 23.53                30.43   
9                    Burnley                 24.24                37.50   
10  Brighton and Hove Albion                 14.71                30.43   
11           West Ham United                 26.47                43.48   
12                   Arsenal                 28.12                48.00   
13         Manchester United                 39.39                66.67   
14                   Everton                 23.53                56.52   

    changes  rate_of_change_%  
0    -25.76            -34.00  
1    -10.23            -25.96  
2     -4.17            -12.50  
3     -2.05             -4.97  
4     -1.14             -3.75  
5      1.52              3.13  
6      1.92              2.84  
7      3.41              8.04  
8      6.91             29.35  
9     13.26             54.69  
10    15.73            106.96  
11    17.01             64.25  
12    19.88             70.67  
13    27.27             69.23  
14    32.99            140.22

Information obtained from home_win_rate dataframe:
- 8 out of 15  clubs had increased win rate significantly with average of 17.3% rate_of_change after crowd suspension when playing at away ground
- Liverpool had dropped win rate significantly both in home and away ground.
- Manchester United and Everton are the teams who had the most increased win rate at away ground after crowd suspension.

##### 2. Points per Game

In [20]:
home_ppg = home_stats('home_ppg', stayed_teams)
home_ppg

Before attendance prohibition, the average home_ppg is 1.74.
After attendance prohibition, the average home_ppg is 1.54.
The average home_ppg changes is -0.21.
The average home_ppg rate_of_change is -12.07%.
There are 11 clubs are negatively affected, 2 relatively unaffected, and 2 positively affected.


home_team  home_ppg_before  home_ppg_after  changes  \
0                  Liverpool             2.88            1.87    -1.01   
1                    Everton             1.79            1.17    -0.62   
2                    Arsenal             2.09            1.65    -0.44   
3                    Burnley             1.35            1.00    -0.35   
4            Manchester City             2.59            2.36    -0.23   
5           Newcastle United             1.39            1.17    -0.23   
6          Manchester United             1.91            1.70    -0.22   
7   Brighton and Hove Albion             1.24            1.04    -0.20   
8             Crystal Palace             1.24            1.04    -0.19   
9    Wolverhampton Wanderers             1.65            1.48    -0.17   
10            Leicester City             1.68            1.52    -0.15   
11         Tottenham Hotspur             1.94            1.92    -0.02   
12                   Chelsea             1.94            1.96     0.02   
13           West Ham United             1.39            1.71     0.31   
14               Southampton             1.09            1.48     0.39   

    rate_of_change_%  
0             -35.14  
1             -34.75  
2             -20.88  
3             -26.09  
4              -9.01  
5             -16.30  
6             -11.30  
7             -16.16  
8             -15.53  
9             -10.25  
10             -9.23  
11             -1.17  
12              0.79  
13             22.55  
14             35.84

Information obtained from home_ppg dataframe:
- 11 out of 15 (~73.33%) clubs had decreased points significantly per game, with average of -12.07% rate_of_change  after crowd suspension when playing at home ground
- Liverpool suffers the most from crowd suspension, having decreased 1.01 points per game
- Tottenham Hotspur and Chelsea home points per game did not get affected by crowd suspension with only slightly changes 

In [21]:
away_ppg = away_stats('away_ppg', stayed_teams)
away_ppg

Before attendance prohibition, the average away_ppg is 1.34.
After attendance prohibition, the average away_ppg is 1.54.
The average away_ppg changes is 0.19.
The average away_ppg rate_of_change is 14.18%.
There are 8 clubs are positively affected, 4 relatively unaffected, and 3 negatively affected.


away_team  away_ppg_before  away_ppg_after  changes  \
0                  Liverpool             2.45            1.79    -0.66   
1             Crystal Palace             1.39            1.00    -0.39   
2    Wolverhampton Wanderers             1.33            1.12    -0.21   
3                    Chelsea             1.64            1.67     0.03   
4                Southampton             1.09            1.12     0.03   
5          Tottenham Hotspur             1.41            1.48     0.07   
6            Manchester City             2.12            2.22     0.10   
7           Newcastle United             1.00            1.13     0.13   
8             Leicester City             1.45            1.67     0.21   
9                    Burnley             1.00            1.29     0.29   
10                   Arsenal             1.22            1.56     0.34   
11  Brighton and Hove Albion             0.71            1.22     0.51   
12           West Ham United             0.97            1.57     0.59   
13                   Everton             0.94            1.87     0.93   
14         Manchester United             1.39            2.33     0.94   

    rate_of_change_%  
0             -27.01  
1             -28.26  
2             -15.62  
3               1.85  
4               3.13  
5               4.71  
6               4.71  
7              13.04  
8              14.58  
9              29.17  
10             28.00  
11             72.46  
12             61.26  
13             98.64  
14             67.39

Information obtained from away_ppg dataframe:
- The crowd suspension effect is parallel.  As a result of dropping points per game for home team, 12 out of 15 clubs had increased points per game (8 clubs increased significantly), with average_rate_of_change 21.87% after crowd suspension when playing at away ground
- Unlike majority of clubs that got this increasing away ppg advantage, 3 clubs; Liverpool, Wolverhampton Wanderers and Crystal Palace  were having decreased away ppg with Liverpool had the most decreased ppg (-0.66)
- Manchester United gained the most points per game (+0.94) from this away ppg advantage.

#### PERFORMANCE STATS

#####  1. Goals per game

In [22]:
#Goals per game (gpg) =average team goals in a match
home_gpg = home_stats('home_goals', stayed_teams)
home_gpg

Before attendance prohibition, the average home_goals is 1.66.
After attendance prohibition, the average home_goals is 1.53.
The average home_goals changes is -0.13.
The average home_goals rate_of_change is -7.83%.
There are 6 clubs are negatively affected, 5 positively affected, and 4 relatively unaffected.


home_team  home_goals_before  home_goals_after  changes  \
0                  Liverpool               2.79              1.78    -1.01   
1                    Arsenal               2.00              1.48    -0.52   
2                    Burnley               1.29              0.78    -0.51   
3                    Everton               1.48              1.21    -0.28   
4    Wolverhampton Wanderers               1.44              1.17    -0.27   
5            Manchester City               2.81              2.68    -0.13   
6                    Chelsea               1.79              1.70    -0.10   
7   Brighton and Hove Albion               1.09              1.04    -0.05   
8            West Ham United               1.67              1.62    -0.04   
9          Tottenham Hotspur               1.85              1.83    -0.02   
10            Crystal Palace               0.91              1.00     0.09   
11            Leicester City               1.59              1.70     0.11   
12               Southampton               1.26              1.43     0.17   
13         Manchester United               1.82              2.13     0.31   
14          Newcastle United               1.09              1.42     0.33   

    rate_of_change_%  
0             -36.20  
1             -26.09  
2             -39.53  
3             -18.62  
4             -18.54  
5              -4.71  
6              -5.49  
7              -4.51  
8              -2.50  
9              -0.82  
10              9.68  
11              6.76  
12             13.45  
13             16.83  
14             29.86

Information obtained from home_gpg dataframe:
- 6 out of 15  clubs significantly had decreased goals per game, with average of -7.83% rate_of_change  after crowd suspension when playing at home ground
- Liverpool had the most dropped goals per game when playing at home (-1.01 goals per game) after crowd suspension

In [23]:
away_gpg = away_stats('away_goals', stayed_teams)
away_gpg

Before attendance prohibition, the average away_goals is 1.36.
After attendance prohibition, the average away_goals is 1.4.
The average away_goals changes is 0.04.
The average away_goals rate_of_change is 2.94%.
There are 7 clubs are positively affected, 5 negatively affected, and 3 relatively unaffected.


away_team  away_goals_before  away_goals_after  changes  \
0             Crystal Palace               1.39              0.96    -0.44   
1    Wolverhampton Wanderers               1.18              0.79    -0.39   
2                    Everton               1.24              1.09    -0.15   
3             Leicester City               1.67              1.58    -0.08   
4                    Chelsea               1.61              1.54    -0.06   
5                    Burnley               1.06              1.00    -0.06   
6            Manchester City               2.15              2.17     0.03   
7   Brighton and Hove Albion               0.91              0.96     0.04   
8                    Arsenal               1.41              1.48     0.07   
9          Tottenham Hotspur               1.56              1.65     0.09   
10                 Liverpool               1.82              1.92     0.10   
11               Southampton               1.12              1.25     0.13   
12          Newcastle United               0.91              1.09     0.18   
13         Manchester United               1.42              1.92     0.49   
14           West Ham United               0.94              1.61     0.67   

    rate_of_change_%  
0             -31.25  
1             -33.01  
2             -12.01  
3              -5.00  
4              -4.01  
5              -5.71  
6               1.25  
7               4.91  
8               5.24  
9               5.99  
10              5.42  
11             11.49  
12             19.21  
13             34.57  
14             70.92

Information obtained from away_gpg dataframe:
- 7 out of 15  clubs significantly had increased goals per game, with average of 4.53% rate_of_change  after crowd suspension when playing at away ground
- West Ham United had the most increased goals per game when playing at away (+0.67 goals per game) after crowd suspension

##### 2. Shots

In [24]:
home_shots = home_stats('home_shots', stayed_teams)
home_shots

Before attendance prohibition, the average home_shots is 14.56.
After attendance prohibition, the average home_shots is 13.06.
The average home_shots changes is -1.5.
The average home_shots rate_of_change is -10.3%.
There are 13 clubs are negatively affected, 1 relatively unaffected, and 1 positively affected.


home_team  home_shots_before  home_shots_after  changes  \
0             Crystal Palace              13.79              9.48    -4.32   
1                    Everton              14.61             11.08    -3.52   
2            Manchester City              20.81             17.68    -3.13   
3          Tottenham Hotspur              15.00             12.17    -2.83   
4                    Chelsea              17.29             15.30    -1.99   
5             Leicester City              14.94             13.00    -1.94   
6                    Burnley              11.50             10.30    -1.20   
7          Manchester United              15.62             14.43    -1.18   
8           Newcastle United              13.36             12.21    -1.16   
9                    Arsenal              13.15             12.17    -0.97   
10                 Liverpool              17.35             16.39    -0.96   
11           West Ham United              12.36             11.58    -0.78   
12               Southampton              13.03             12.30    -0.73   
13   Wolverhampton Wanderers              13.53             13.70     0.17   
14  Brighton and Hove Albion              12.06             14.08     2.02   

    rate_of_change_%  
0             -31.29  
1             -24.12  
2             -15.05  
3             -18.89  
4             -11.51  
5             -12.99  
6             -10.40  
7              -7.57  
8              -8.65  
9              -7.40  
10             -5.54  
11             -6.31  
12             -5.56  
13              1.23  
14             16.77

In [25]:
away_shots = away_stats('away_shots', stayed_teams)
away_shots

Before attendance prohibition, the average away_shots is 11.66.
After attendance prohibition, the average away_shots is 11.67.
The average away_shots changes is 0.01.
The average away_shots rate_of_change is 0.09%.
There are 6 clubs are relatively unaffected, 5 positively affected, and 4 negatively affected.


away_team  away_shots_before  away_shots_after  changes  \
0                    Everton              11.85              9.39    -2.46   
1    Wolverhampton Wanderers              11.79              9.88    -1.91   
2                Southampton              12.45             11.21    -1.25   
3          Tottenham Hotspur              11.59             10.57    -1.02   
4            Manchester City              16.41             15.70    -0.72   
5                    Chelsea              15.15             14.50    -0.65   
6             Crystal Palace               9.42              9.17    -0.26   
7           Newcastle United               8.97              8.83    -0.14   
8   Brighton and Hove Albion              10.26             10.22    -0.05   
9          Manchester United              12.85             12.88     0.03   
10                   Arsenal              10.38             10.92     0.54   
11            Leicester City              12.48             13.38     0.89   
12                   Burnley               8.18              9.71     1.53   
13                 Liverpool              13.24             15.42     2.17   
14           West Ham United               9.82             13.30     3.48   

    rate_of_change_%  
0             -20.77  
1             -16.23  
2             -10.01  
3              -8.83  
4              -4.36  
5              -4.30  
6              -2.73  
7              -1.61  
8              -0.46  
9               0.21  
10              5.25  
11              7.13  
12             18.66  
13             16.42  
14             35.43

##### 3. Shots on Target

In [26]:
home_SoT = home_stats('home_SoT', stayed_teams)
home_SoT

Before attendance prohibition, the average home_SoT is 5.02.
After attendance prohibition, the average home_SoT is 4.7.
The average home_SoT changes is -0.32.
The average home_SoT rate_of_change is -6.37%.
There are 9 clubs are negatively affected, 3 relatively unaffected, and 3 positively affected.


home_team  home_SoT_before  home_SoT_after  changes  \
0                    Everton             4.94            3.79    -1.15   
1            West Ham United             4.88            4.04    -0.84   
2                  Liverpool             6.71            5.91    -0.79   
3          Manchester United             6.47            5.70    -0.77   
4                    Burnley             4.00            3.26    -0.74   
5            Manchester City             7.38            6.68    -0.70   
6                    Arsenal             4.88            4.30    -0.58   
7             Crystal Palace             3.82            3.26    -0.56   
8             Leicester City             5.06            4.78    -0.28   
9                    Chelsea             6.21            6.13    -0.08   
10          Newcastle United             4.06            4.25     0.19   
11         Tottenham Hotspur             4.94            5.17     0.23   
12   Wolverhampton Wanderers             4.18            4.52     0.35   
13               Southampton             4.38            4.83     0.44   
14  Brighton and Hove Albion             3.33            3.83     0.50   

    rate_of_change_%  
0             -23.24  
1             -17.16  
2             -11.82  
3             -11.98  
4             -18.48  
5              -9.42  
6             -11.84  
7             -14.72  
8              -5.46  
9              -1.22  
10              4.66  
11              4.60  
12              8.27  
13             10.13  
14             15.00

In [27]:
away_SoT = away_stats('away_SoT', stayed_teams)
away_SoT

Before attendance prohibition, the average away_SoT is 4.22.
After attendance prohibition, the average away_SoT is 4.2.
The average away_SoT changes is -0.01.
The average away_SoT rate_of_change is -0.24%.
There are 7 clubs are positively affected, 6 negatively affected, and 2 relatively unaffected.


away_team  away_SoT_before  away_SoT_after  changes  \
0          Tottenham Hotspur             4.59            3.57    -1.02   
1            Manchester City             6.35            5.43    -0.92   
2    Wolverhampton Wanderers             3.88            3.33    -0.55   
3                    Everton             4.15            3.65    -0.49   
4             Leicester City             5.03            4.67    -0.36   
5   Brighton and Hove Albion             3.50            3.17    -0.33   
6             Crystal Palace             3.39            3.46     0.06   
7                Southampton             4.36            4.46     0.09   
8           Newcastle United             3.09            3.30     0.22   
9                    Arsenal             3.53            3.80     0.27   
10                 Liverpool             5.27            5.54     0.27   
11         Manchester United             5.18            5.50     0.32   
12                   Chelsea             4.73            5.12     0.40   
13           West Ham United             3.65            4.48     0.83   
14                   Burnley             2.55            3.54     1.00   

    rate_of_change_%  
0             -22.30  
1             -14.45  
2             -14.06  
3             -11.93  
4              -7.23  
5              -9.32  
6               1.90  
7               2.17  
8               7.00  
9               7.61  
10              5.10  
11              6.14  
12              8.41  
13             22.79  
14             39.14

##### 4. Shots on Target Rate

In [28]:
home_SoT_rate = home_stats('home_SoT_rate', stayed_teams)
home_SoT_rate

Before attendance prohibition, the average home_SoT_rate is 34.94.
After attendance prohibition, the average home_SoT_rate is 37.04.
The average home_SoT_rate changes is 2.1.
The average home_SoT_rate rate_of_change is 6.01%.
There are 8 clubs are positively affected, 5 relatively unaffected, and 2 negatively affected.


home_team  home_SoT_rate_before  home_SoT_rate_after  \
0          Manchester United                 44.37                37.55   
1                    Burnley                 34.64                31.61   
2                  Liverpool                 39.05                37.29   
3            West Ham United                 39.46                37.92   
4                    Everton                 34.06                33.68   
5   Brighton and Hove Albion                 27.46                28.17   
6             Leicester City                 35.16                36.32   
7                    Arsenal                 37.01                39.07   
8    Wolverhampton Wanderers                 31.02                33.19   
9            Manchester City                 35.74                38.94   
10               Southampton                 36.56                40.81   
11                   Chelsea                 36.39                41.24   
12          Newcastle United                 30.58                36.46   
13         Tottenham Hotspur                 35.27                42.51   
14            Crystal Palace                 27.36                40.80   

    changes  rate_of_change_%  
0     -6.82            -15.36  
1     -3.03             -8.74  
2     -1.76             -4.51  
3     -1.54             -3.90  
4     -0.38             -1.13  
5      0.71              2.58  
6      1.16              3.29  
7      2.06              5.56  
8      2.17              7.00  
9      3.20              8.95  
10     4.25             11.62  
11     4.84             13.30  
12     5.88             19.23  
13     7.25             20.55  
14    13.45             49.14

Information obtained from home_SoT_rate (shots on target percentage of home team) dataframe:
- 8 out of 15  clubs significantly had increased shots on target percentage, with average of 6.01% rate_of_change  after crowd suspension when playing at home ground
- 2 teams who had the most significant changes of SoT percentage at home are Crystal Palace and Manchester United. Crystal Palace SoT percentage significantly increased and Manchester United SoT percentage significantly dropped.

In [29]:
away_SoT_rate= away_stats('away_SoT_rate', stayed_teams)
away_SoT_rate

Before attendance prohibition, the average away_SoT_rate is 36.51.
After attendance prohibition, the average away_SoT_rate is 36.54.
The average away_SoT_rate changes is 0.03.
The average away_SoT_rate rate_of_change is 0.08%.
There are 6 clubs are negatively affected, 6 positively affected, and 3 relatively unaffected.


away_team  away_SoT_rate_before  away_SoT_rate_after  \
0          Tottenham Hotspur                 40.65                34.87   
1            West Ham United                 37.73                33.06   
2                  Liverpool                 39.90                36.62   
3            Manchester City                 40.14                37.85   
4                    Arsenal                 36.90                34.64   
5             Crystal Palace                 37.69                35.48   
6             Leicester City                 39.59                38.06   
7   Brighton and Hove Albion                 33.80                33.08   
8    Wolverhampton Wanderers                 32.64                32.74   
9                Southampton                 37.03                40.20   
10         Manchester United                 39.96                43.61   
11          Newcastle United                 33.48                37.22   
12                   Chelsea                 32.08                35.87   
13                   Everton                 35.13                39.46   
14                   Burnley                 30.85                35.28   

    changes  rate_of_change_%  
0     -5.78            -14.22  
1     -4.68            -12.39  
2     -3.29             -8.24  
3     -2.30             -5.72  
4     -2.26             -6.12  
5     -2.22             -5.88  
6     -1.53             -3.86  
7     -0.72             -2.12  
8      0.10              0.31  
9      3.17              8.56  
10     3.65              9.14  
11     3.74             11.17  
12     3.78             11.79  
13     4.34             12.35  
14     4.43             14.37

Information obtained from away_SoT_rate (shots on target percentage of away team) dataframe:
- 2 teams who had the most significant changes of SoT percentage at away are Burnley and Tottenham Hotspur. Burnley SoT percentage significantly increased and Tottenham Hotspur SoT percentage significantly dropped

##### 5. Quantity Conversion Rate

Quantity conversion rate is goal conversion rate from all shots created (shots on target + shots off target).

In [30]:
home_QuanCR = home_stats('home_QuanCR', stayed_teams)
home_QuanCR

Before attendance prohibition, the average home_QuanCR is 12.04.
After attendance prohibition, the average home_QuanCR is 13.21.
The average home_QuanCR changes is 1.16.
The average home_QuanCR rate_of_change is 9.63%.
There are 9 clubs are positively affected, 5 negatively affected, and 1 relatively unaffected.


home_team  home_QuanCR_before  home_QuanCR_after  changes  \
0                  Liverpool               16.65              12.62    -4.03   
1    Wolverhampton Wanderers               12.47               8.64    -3.84   
2                    Burnley               11.96               8.14    -3.82   
3   Brighton and Hove Albion               10.58               8.50    -2.08   
4                    Arsenal               15.34              14.33    -1.01   
5                    Chelsea               10.69              11.13     0.44   
6          Manchester United               13.12              13.98     0.86   
7             Leicester City               11.52              12.54     1.01   
8                    Everton               10.22              11.94     1.71   
9            Manchester City               14.21              16.17     1.95   
10               Southampton               10.54              12.73     2.19   
11         Tottenham Hotspur               14.61              17.66     3.04   
12          Newcastle United                9.47              13.46     3.99   
13           West Ham United               12.68              17.62     4.94   
14            Crystal Palace                6.59              18.67    12.09   

    rate_of_change_%  
0             -24.20  
1             -30.77  
2             -31.96  
3             -19.67  
4              -6.58  
5               4.13  
6               6.52  
7               8.80  
8              16.75  
9              13.75  
10             20.79  
11             20.84  
12             42.12  
13             38.93  
14            183.57

Information obtained from home_QuanCR (quantity conversion rate of home team) dataframe:

- 9 out of 15  clubs significantly had increased quantity goal conversion rate, with average of 9.63% rate_of_change  after crowd suspension when playing at home ground.
- Crystal Palace has the most significantly increased quantity conversion rate at home after crowd suspension (12.09%)
- Liverpool has the most significantly dropped quantity conversion rate at home after crowd suspension (-4.03%)

In [31]:
away_QuanCR = away_stats('away_QuanCR', stayed_teams)
away_QuanCR

Before attendance prohibition, the average away_QuanCR is 12.75.
After attendance prohibition, the average away_QuanCR is 12.28.
The average away_QuanCR changes is -0.47.
The average away_QuanCR rate_of_change is -3.69%.
There are 7 clubs are positively affected, 5 negatively affected, and 3 relatively unaffected.


away_team  away_QuanCR_before  away_QuanCR_after  changes  \
0                    Burnley               16.21               9.09    -7.11   
1             Crystal Palace               17.15              11.49    -5.66   
2    Wolverhampton Wanderers               11.97               7.75    -4.22   
3                    Arsenal               16.29              14.22    -2.08   
4                  Liverpool               14.07              12.55    -1.51   
5                    Chelsea               10.90              10.99     0.09   
6           Newcastle United               10.09              10.36     0.27   
7          Tottenham Hotspur               14.91              15.44     0.53   
8   Brighton and Hove Albion                9.82              10.42     0.60   
9             Leicester City               13.87              14.79     0.92   
10               Southampton               10.14              11.13     0.99   
11           Manchester City               13.88              15.65     1.77   
12           West Ham United                9.65              11.47     1.82   
13                   Everton               10.49              13.05     2.55   
14         Manchester United               11.76              15.73     3.97   

    rate_of_change_%  
0             -43.88  
1             -32.98  
2             -35.25  
3             -12.75  
4             -10.75  
5               0.81  
6               2.68  
7               3.55  
8               6.11  
9               6.66  
10              9.80  
11             12.74  
12             18.87  
13             24.35  
14             33.72

Information obtained from away_QuanCR (quantity conversion rate of away team) dataframe:
- Manchester United has the most significantly increased quantity conversion rate at away after crowd suspension
- Burnley has the most significantly dropped quantity conversion rate at home after crowd suspension

##### 6. Quality Conversion Rate

Quality conversion rate is goal conversion rate from shots on target (excluding shots off target).

In [32]:
home_QualCR = home_stats('home_QualCR', stayed_teams)
home_QualCR 

Before attendance prohibition, the average home_QualCR is 34.03.
After attendance prohibition, the average home_QualCR is 33.17.
The average home_QualCR changes is -0.86.
The average home_QualCR rate_of_change is -2.53%.
There are 7 clubs are positively affected, 6 negatively affected, and 2 relatively unaffected.


home_team  home_QualCR_before  home_QualCR_after  changes  \
0                    Burnley               39.20              23.34   -15.85   
1                  Liverpool               43.14              29.11   -14.03   
2    Wolverhampton Wanderers               38.27              27.72   -10.55   
3                    Arsenal               42.15              32.54    -9.62   
4   Brighton and Hove Albion               32.88              26.25    -6.63   
5                    Chelsea               32.38              27.25    -5.13   
6                Southampton               30.21              30.14    -0.07   
7             Leicester City               31.82              32.37     0.55   
8           Newcastle United               29.05              30.61     1.56   
9            Manchester City               38.18              40.66     2.48   
10         Tottenham Hotspur               37.67              41.83     4.16   
11         Manchester United               27.96              35.87     7.92   
12                   Everton               31.84              40.60     8.76   
13           West Ham United               31.87              41.57     9.69   
14            Crystal Palace               23.80              37.64    13.85   

    rate_of_change_%  
0             -40.45  
1             -32.53  
2             -27.56  
3             -22.81  
4             -20.16  
5             -15.84  
6              -0.22  
7               1.72  
8               5.35  
9               6.49  
10             11.05  
11             28.32  
12             27.51  
13             30.42  
14             58.18

Information obtained from home_QualCR (quality conversion rate of home team) dataframe:
- Crystal Palace has the most significantly increased quality conversion rate at home after crowd suspension 
- Burnley and Liverpool have the most significantly dropped quality conversion rate at home after crowd suspension 

In [33]:
away_QualCR = away_stats('away_QualCR', stayed_teams)
away_QualCR

Before attendance prohibition, the average away_QualCR is 32.3.
After attendance prohibition, the average away_QualCR is 31.55.
The average away_QualCR changes is -0.74.
The average away_QualCR rate_of_change is -2.29%.
There are 7 clubs are positively affected, 6 negatively affected, and 2 relatively unaffected.


away_team  away_QualCR_before  away_QualCR_after  changes  \
0                    Burnley               37.83              23.28   -14.54   
1             Crystal Palace               42.34              28.53   -13.81   
2    Wolverhampton Wanderers               32.07              23.06    -9.02   
3                    Arsenal               40.90              34.65    -6.26   
4                    Chelsea               35.21              30.73    -4.48   
5                  Liverpool               34.80              32.88    -1.92   
6           Newcastle United               30.13              29.38    -0.75   
7             Leicester City               34.25              35.37     1.12   
8                    Everton               28.04              29.83     1.80   
9                Southampton               25.36              28.67     3.31   
10         Manchester United               29.93              34.70     4.77   
11         Tottenham Hotspur               36.28              41.16     4.88   
12  Brighton and Hove Albion               22.79              30.07     7.29   
13           Manchester City               32.87              40.71     7.84   
14           West Ham United               21.63              30.23     8.60   

    rate_of_change_%  
0             -38.45  
1             -32.62  
2             -28.11  
3             -15.30  
4             -12.72  
5              -5.51  
6              -2.48  
7               3.28  
8               6.40  
9              13.07  
10             15.95  
11             13.45  
12             31.98  
13             23.86  
14             39.75

Information obtained from home_QualCR (quality conversion rate of home team) dataframe:
- West Ham has the most significantly increased quality conversion rate at away after crowd suspension 
- Burnley and Crystal Palace have the most significantly dropped quality conversion rate at away after crowd suspension 

##### 7. Possession

In [34]:
home_poss = home_stats('home_possession', stayed_teams)
home_poss

Before attendance prohibition, the average home_possession is 53.11.
After attendance prohibition, the average home_possession is 52.43.
The average home_possession changes is -0.68.
The average home_possession rate_of_change is -1.28%.
There are 7 clubs are relatively unaffected, 5 negatively affected, and 3 positively affected.


home_team  home_possession_before  home_possession_after  \
0          Tottenham Hotspur                   58.73                  50.82   
1                    Arsenal                   59.00                  52.74   
2                    Everton                   52.55                  47.93   
3             Crystal Palace                   45.98                  41.80   
4            West Ham United                   47.63                  43.60   
5                    Chelsea                   62.71                  60.49   
6            Manchester City                   67.76                  66.72   
7             Leicester City                   55.70                  55.65   
8           Newcastle United                   41.60                  41.93   
9   Brighton and Hove Albion                   49.81                  51.43   
10                 Liverpool                   62.85                  64.78   
11                   Burnley                   42.55                  44.67   
12         Manchester United                   55.76                  58.85   
13   Wolverhampton Wanderers                   48.48                  52.39   
14               Southampton                   45.49                  52.62   

    changes  rate_of_change_%  
0     -7.91            -13.47  
1     -6.26            -10.60  
2     -4.63             -8.80  
3     -4.18             -9.09  
4     -4.03             -8.45  
5     -2.21             -3.53  
6     -1.04             -1.54  
7     -0.04             -0.08  
8      0.33              0.80  
9      1.61              3.24  
10     1.93              3.06  
11     2.12              4.98  
12     3.09              5.54  
13     3.91              8.07  
14     7.13             15.67

In [35]:
away_stats('away_possession', stayed_teams)

Before attendance prohibition, the average away_possession is 50.21.
After attendance prohibition, the average away_possession is 50.26.
The average away_possession changes is 0.05.
The average away_possession rate_of_change is 0.1%.
There are 7 clubs are relatively unaffected, 4 negatively affected, and 4 positively affected.


away_team  away_possession_before  away_possession_after  \
0                    Everton                   49.26                  45.14   
1            Manchester City                   66.74                  62.74   
2            West Ham United                   45.55                  42.45   
3             Crystal Palace                   43.43                  40.45   
4                    Arsenal                   55.41                  52.71   
5          Tottenham Hotspur                   54.25                  51.94   
6                  Liverpool                   62.07                  61.16   
7   Brighton and Hove Albion                   46.15                  46.90   
8           Newcastle United                   35.99                  36.77   
9          Manchester United                   53.54                  54.75   
10                   Chelsea                   61.21                  62.85   
11                   Burnley                   38.42                  40.45   
12   Wolverhampton Wanderers                   45.12                  47.54   
13            Leicester City                   51.19                  55.57   
14               Southampton                   44.78                  52.43   

    changes  rate_of_change_%  
0     -4.12             -8.37  
1     -4.00             -6.00  
2     -3.10             -6.81  
3     -2.98             -6.86  
4     -2.69             -4.86  
5     -2.31             -4.26  
6     -0.91             -1.46  
7      0.75              1.62  
8      0.78              2.18  
9      1.22              2.27  
10     1.64              2.67  
11     2.03              5.28  
12     2.43              5.38  
13     4.37              8.54  
14     7.65             17.08

Conclusions and insights from stats comparison before and after attendance prohibition:
1. For results stats, we can conclude straight-forwardly that home result decreases and followed by increases of away result.
2. For performance stats, in general home performance became worse in quantity but better in quality. Away performs slightly better after covid-19 while some stats relatively unaffected.
3. Generally home stats became worse after covid-19 interruption. But interestingly, this does not translates to better away stats, which means the absence of the home crowd heavily impacted home performance but not necessarily made away team performs better.
4. While number of home shots and shots on target have decreased, the quality of the shot increased as reflected on increased SoT%, Quan_CR and Qual_CR
5. Attendance prohibition did not affect a lot on ball possession 

## Visualization

In [36]:
'''def home_stat_viz(stat, team, summary):
    fig = go.Figure(data=[
        go.Bar(name= stat+' before crowd suspension', 
               x=home_stats(stat,team, summary)['home_team'], 
               y=home_stats(stat,team, summary)[stat+'_before'],
               marker_color='mediumspringgreen',
               text=home_stats(stat,team, summary)[stat+'_before'], 
               textposition='auto',
               texttemplate='%{text:.3s}'),
         go.Bar(name=stat+' after crowd suspension', 
                x=home_stats(stat,team, summary)['home_team'], 
                y=home_stats(stat,team, summary)[stat+'_after'],
                marker_color='lightsalmon',
                text=home_stats(stat,team, summary)[stat+'_after'],
                textposition='auto',
                texttemplate='%{text:.2f}')
    ])
    # Customize layour
    fig.update_layout(barmode='group',
                      title={
                            'text': '{} of Premier League Clubs Before and After Crowd Suspension'.format(stat),
                            'y':0.9,
                            'x':0,
                            'xanchor': 'left',
                            'yanchor': 'top'},
                      xaxis=dict(
                                    showline=True,
                                    showgrid=False,
                                    showticklabels=True,
                                    linewidth=2,
                                    ticks='outside',
                                    tickfont=dict(family='Arial',
                                                  size=12
                                    ),
                                ),
                        yaxis=dict(
                                        showgrid=False,
                                        zeroline=True,
                                        showline=True,
                                        showticklabels=True,
                                    ),
                        autosize=True,
                        margin=dict(
                                        autoexpand=True,
                                        l=0,
                                        r=0,
                                        t=80,
                                    ),
                        plot_bgcolor='azure')
    fig.show()'''

"def home_stat_viz(stat, team, summary):\n    fig = go.Figure(data=[\n        go.Bar(name= stat+' before crowd suspension', \n               x=home_stats(stat,team, summary)['home_team'], \n               y=home_stats(stat,team, summary)[stat+'_before'],\n               marker_color='mediumspringgreen',\n               text=home_stats(stat,team, summary)[stat+'_before'], \n               textposition='auto',\n               texttemplate='%{text:.3s}'),\n         go.Bar(name=stat+' after crowd suspension', \n                x=home_stats(stat,team, summary)['home_team'], \n                y=home_stats(stat,team, summary)[stat+'_after'],\n                marker_color='lightsalmon',\n                text=home_stats(stat,team, summary)[stat+'_after'],\n                textposition='auto',\n                texttemplate='%{text:.2f}')\n    ])\n    # Customize layour\n    fig.update_layout(barmode='group',\n                      title={\n                            'text': '{} of Premier Leagu

In [53]:
def home_stat_change(stat, team, summary):
    #Labels color
    color = np.where(home_stats(stat,team, summary)['changes']>0,
                                                                   'mediumspringgreen','tomato')
    fig = go.Figure(data=[
        go.Bar(x=home_stats(stat,team, summary)['home_team'], 
               y=home_stats(stat,team, summary)['changes'],
               marker_color=color,
               text=home_stats(stat,team, summary)['changes'], 
               textposition='auto',
               texttemplate='%{text:.2f}'),
    ])
    # Customize layour
    fig.update_layout(barmode='group',
                      title={
                            'text': '{} changes of Premier League Clubs Before and After Crowd Suspension'.format(stat),
                            'y':0.9,
                            'x':0.5,
                            'xanchor': 'center',
                            'yanchor': 'top'},
                      xaxis=dict(
                                    showline=True,
                                    showgrid=False,
                                    showticklabels=True,
                                    linewidth=2,
                                    ticks='outside',
                                    tickfont=dict(family='Arial',
                                                  size=12
                                    ),
                                ),
                        yaxis=dict(
                                        showgrid=False,
                                        zeroline=True,
                                        showline=True,
                                        showticklabels=True,
                                        title=stat
                                    ),
                        autosize=True,
                        margin=dict(
                                        autoexpand=True,
                                        l=0,
                                        r=0,
                                        t=80,
                                    ),
                        plot_bgcolor='azure')
    fig.show()

In [51]:
def away_stat_change(stat, team, summary):
    #Labels color
    color = np.where(away_stats(stat,team, summary)['changes']>0,
                                                                   'mediumspringgreen','tomato')
    fig = go.Figure(data=[
        go.Bar(x=away_stats(stat,team, summary)['away_team'], 
               y=away_stats(stat,team, summary)['changes'],
               marker_color=color,
               text=away_stats(stat,team, summary)['changes'], 
               textposition='auto',
               texttemplate='%{text:.2f}'),
    ])
    # Customize layour
    fig.update_layout(barmode='group',
                      title={
                            'text': '{} changes of Premier League Clubs Before and After Crowd Suspension'.format(stat),
                            'y':0.9,
                            'x':0.5,
                            'xanchor': 'center',
                            'yanchor': 'top'},
                        autosize=True,
                        margin=dict(
                                        autoexpand=True,
                                        l=0,
                                        r=0,
                                        t=80,
                                    ),
                        plot_bgcolor='azure')
    fig.show()

### Result Stats Visualization

#### Home Stats

In [54]:
home_stat_change('home_win_rate', stayed_teams, summary=False)

In [40]:
home_stat_change('home_ppg', stayed_teams, summary=False)

**Insights**
<br>
1. From the graphs above shown that Liverpool, Arsenal, Burnley, and Everton are the teams who had decreased the most in home win rate, home points per game since there was no spectators. 
2. Tottenham and Chelsea are the 2 teams whose stats did not get affected meaningfully by no spectators when playing at home stadium.

#### Away Stats

In [41]:
away_stat_change('away_win_rate', stayed_teams, summary=False)

In [42]:
away_stat_change('away_ppg', stayed_teams, summary=False)

**Insights**
<br>
1. Manchester United and Everton got the best results  when playing at away ground after covid-19 interruption
2. Interestingly, like in home game, Liverpool also had the most dropped points per game and win rate in away game.

#### Performance Stats Visualization

#### Home Stats

In [43]:
home_stat_change('home_goals', stayed_teams, summary=False)

In [44]:
home_stat_change('home_shots', stayed_teams, summary=False)

In [45]:
home_stat_change('home_SoT', stayed_teams, summary=False)

In [46]:
home_stat_change('home_SoT_rate', stayed_teams, summary=False)

In [47]:
home_stat_change('home_QuanCR', stayed_teams, summary=False)

In [48]:
home_stat_change('home_QualCR', stayed_teams, summary=False)

**Insights**
<br>
1. Liverpool, Arsenal, Burnley, and Everton are the teams who really had hard time to score goals at their home after covid-19 prohibition. They have the most decreased goals per game in home game. Their inability to scored goals at home costed them dropped home win rate and points per game.
2. Everton had the most decreased home shots and home shots on target. Those maybe that causes their decreased goals at home.
3. Manchester United home goals were increasing despite they had the most decreased shots on target rate.
4. Liverpool, Burnley, and Wolverhampton Wanderers had the most decreased quantity and quality goal conversion rate at home after covid-19 interruption. Their ability to convert shots into goals was incredibly getting worse.

#### Away Stats

In [56]:
away_stat_change('away_goals', stayed_teams, summary=False)

In [57]:
away_stat_change('away_shots', stayed_teams, summary=False)

In [58]:
away_stat_change('away_SoT', stayed_teams, summary=False)

In [59]:
away_stat_change('away_SoT_rate', stayed_teams, summary=False)

In [60]:
away_stat_change('away_QuanCR', stayed_teams, summary=False)

In [61]:
away_stat_change('away_QualCR', stayed_teams, summary=False)

- Everton's goals per game at away game were decreasing, however they were one of the team who got the most increased points per game and win rate at away game. They scored fewer goals but still managed to grab more points at away game.
- Among other teams, West Ham United have the most increased away goals per game, away quality goal conversion rate, and away shots after covid-19 interruption